### PR TITLE
net/lora : Expose configurations to Kconfig

### PR DESCRIPTION
**Pull request by akshaim**
### Contribution description - 

This PR exposes compile configurations in **net/lora** to Kconfig.

Skipped `LORA_FREQUENCY_RESOLUTION_DEFAULT` since it was planned to be removed. #13318

### Testing procedure

1. New documentation was built using Doxygen 

   The build works fine.

  2. Test files were added to tests/net_lora

      The test file can be found [here](https://github.com/akshaim/RIOT/commit/2336b333eb3d676e8a659c7927b16f3e3f980501)

      Compiled binaries for b-l072z-lrwan1 and tested on the evaluation board. 

3. Tested gnrc_lorawan example in `examples 
`
##### Firmware Output
```
main(): This is RIOT! (Version: 2020.10-devel-1308-gae087-Kconfig_lora_tests)
CONFIG_LORA_PREAMBLE_LENGTH_DEFAULT=(8U)
CONFIG_LORA_SYMBOL_TIMEOUT_DEFAULT=(10U)
CONFIG_LORA_BW_DEFAULT=(LORA_BW_125_KHZ)
CONFIG_LORA_SF_DEFAULT=(LORA_SF7)
CONFIG_LORA_CR_DEFAULT=(LORA_CR_4_5)
CONFIG_LORA_IQ_INVERTED_DEFAULT=CONFIG_LORA_IQ_INVERTED_DEFAULT
CONFIG_LORA_FREQUENCY_HOPPING_DEFAULT=CONFIG_LORA_FREQUENCY_HOPPING_DEFAULT
CONFIG_LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT=(0U)
CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT=CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT
CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT=CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT
LORA_PAYLOAD_CRC_ON_DEFAULT=(true)
CONFIG_LORA_PAYLOAD_LENGTH_DEFAULT=(0U)

```
#### Usage with CFLAGS :
```
CFLAGS += -DCONFIG_LORA_PREAMBLE_LENGTH_DEFAULT=10
CFLAGS += -DCONFIG_LORA_SYMBOL_TIMEOUT_DEFAULT=20
CFLAGS += -DCONFIG_LORA_BW_DEFAULT=LORA_BW_250_KHZ
CFLAGS += -DCONFIG_LORA_SF_DEFAULT=LORA_SF8
CFLAGS += -DCONFIG_LORA_CR_DEFAULT=LORA_CR_4_6
CFLAGS += -DCONFIG_LORA_IQ_INVERTED_DEFAULT=1
CFLAGS += -DCONFIG_LORA_FREQUENCY_HOPPING_DEFAULT=1
CFLAGS += -DCONFIG_LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT=10
# CFLAGS += -DCONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT=1
CFLAGS += -DLORA_PAYLOAD_CRC_ON_DEFAULT=1
CFLAGS += -DCONFIG_LORA_PAYLOAD_LENGTH_DEFAULT=10
```

##### Firmware Output

```
main(): This is RIOT! (Version: 2020.10-devel-1308-gae087-Kconfig_lora_tests)
CONFIG_LORA_PREAMBLE_LENGTH_DEFAULT=10
CONFIG_LORA_SYMBOL_TIMEOUT_DEFAULT=20
CONFIG_LORA_BW_DEFAULT=LORA_BW_250_KHZ
CONFIG_LORA_SF_DEFAULT=LORA_SF8
DCONFIG_LORA_CR_DEFAULT=LORA_CR_4_6
DCONFIG_LORA_IQ_INVERTED_DEFAULT=1
DCONFIG_LORA_FREQUENCY_HOPPING_DEFAULT=1
CONFIG_LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT=10
CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT=CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT
CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT=CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT
CFLAGS += -DLORA_PAYLOAD_CRC_ON_DEFAULT=1
CFLAGS += -DCONFIG_LORA_PAYLOAD_LENGTH_DEFAULT=10
```

#### Usage with menuconfig :

`make menuconfig`

##### Default Firmware Output

```
main(): This is RIOT! (Version: 2020.10-devel-1308-gae087-Kconfig_lora_tests)
CONFIG_LORA_PREAMBLE_LENGTH_DEFAULT=8
CONFIG_LORA_SYMBOL_TIMEOUT_DEFAULT=10
CONFIG_LORA_BW_DEFAULT=(LORA_BW_125_KHZ)
CONFIG_LORA_SF_DEFAULT=(LORA_SF7)
CONFIG_LORA_CR_DEFAULT=(LORA_CR_4_5)
CONFIG_LORA_IQ_INVERTED_DEFAULT=CONFIG_LORA_IQ_INVERTED_DEFAULT
CONFIG_LORA_FREQUENCY_HOPPING_DEFAULT=CONFIG_LORA_FREQUENCY_HOPPING_DEFAULT
CONFIG_LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT=(0U)
CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT=CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT
CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT=CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT
LORA_PAYLOAD_CRC_ON_DEFAULT=(true)
CONFIG_LORA_PAYLOAD_LENGTH_DEFAULT=10
```

##### Configured Firmware Output

```
main(): This is RIOT! (Version: 2020.10-devel-1308-gae087-Kconfig_lora_tests)
CONFIG_LORA_PREAMBLE_LENGTH_DEFAULT=10
CONFIG_LORA_SYMBOL_TIMEOUT_DEFAULT=15
CONFIG_LORA_BW_DEFAULT=(LORA_BW_250_KHZ)
CONFIG_LORA_SF_DEFAULT=(LORA_SF9)
CONFIG_LORA_CR_DEFAULT=(LORA_CR_4_6)
CONFIG_LORA_IQ_INVERTED_DEFAULT=1
CONFIG_LORA_FREQUENCY_HOPPING_DEFAULT=1
CONFIG_LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT=100
CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT=1
CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT=1
LORA_PAYLOAD_CRC_ON_DEFAULT=(false)
CONFIG_LORA_PAYLOAD_LENGTH_DEFAULT=0
```

**MACROS were successfully configured.**

### Issues/PRs references

#12888

### CODE REVIEW

**Comment by akshaim** [comment]
@jia200x  Hey. Appreciate if you can suggest the unit of "LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT". Not sure if it milliseconds or microseconds. The official documentation Semtech does not see to have it.

File: sys/include/net/lora.h [hunk 1]
```diff
@@ -109,9 +109,9 @@ extern "C" {
 #define CONFIG_LORA_IQ_INVERTED_DEFAULT
 #endif
 
-/** @brief Frequency hopping on */
-#ifndef LORA_FREQUENCY_HOPPING_DEFAULT
-#define LORA_FREQUENCY_HOPPING_DEFAULT         (false)
+/** @brief Set this to true to enable frequency hopping */
```
[the remark this answers is not in the document]
Comment (akshaim) [reply sys/include/net/lora.h]: Done

File: sys/include/net/lora.h [hunk 2]
```diff
@@ -119,9 +119,9 @@ extern "C" {
 #define CONFIG_LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT (0U)
 #endif
 
-/** @brief Set fixed header length mode (implicit header) */
-#ifndef LORA_FIXED_HEADER_LEN_MODE_DEFAULT
-#define LORA_FIXED_HEADER_LEN_MODE_DEFAULT     (false)
+/** @brief Set this to true to enable fixed header length mode (implicit header) */
```
[the remark this answers is not in the document]
Comment (akshaim) [reply sys/include/net/lora.h]: Done

File: sys/net/lora/Makefile [hunk 3]
```diff
@@ -0,0 +1 @@
+include $(RIOTBASE)/Makefile.base
```
Comment (leandrolanzieri) [inline sys/net/lora/Makefile:1]: I don't think this is needed...

Comment (akshaim) [reply sys/net/lora/Makefile:1]: I found the same in empty repositories like [these](https://github.com/RIOT-OS/RIOT/tree/master/sys/net/routing) hence added the same.

Comment (leandrolanzieri) [reply sys/net/lora/Makefile:1]: It seems to be forgotten from when NHDP was removed in #13396. It's not needed.

**Comment by jia200x** [comment]
> @jia200x Hey. Appreciate if you can suggest the unit of "LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT". Not sure if it milliseconds or microseconds. The official documentation from Semtech for sx127x chips does not seem to have it.

The Hop Period is expressed in "symbols", not in a time unit :)

File: sys/include/net/lora.h [hunk 4]
```diff
@@ -38,66 +40,161 @@ extern "C" {
...
+#ifndef CONFIG_LORA_CR_DEFAULT
+#define CONFIG_LORA_CR_DEFAULT                      (LORA_CR_4_5)
 #endif
 
-/** @brief Frequency hopping period */
-#ifndef LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT
-#define LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT  (0U)
+/** @brief Set this to 1 to enable inverted I/Q mode
+ *
+ * Inverted I/Q allows differentiating sender and receiver messages. In the case
```
Comment (jia200x) [inline sys/include/net/lora.h:138]: Note that this is a modulation parameter that inverts the I and Q signals. If IQ doesn't match the transceiver is not able to demodulate the LoRa frame.
In case of LoRaWAN, the gateway sends downlink frames using inverted IQ, but all uplink frames are received using non-inverted IQ. But this parameter is not LoRaWAN specific.

Thus, I propose to rephrase this to indicate what the IQ parameter means, and possibly mention the case of LoRaWAN as an example

File: sys/net/lora/Kconfig [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+config LORA_CR_DEFAULT_CR_4_8
+    bool "CR 4/8"
+
+endchoice
+
+config LORA_IQ_INVERTED_DEFAULT
+    bool "Enable inverted I/Q mode"
+    help
+        Enable Inverted I/Q while sending LoRa frames. Inverted I/Q allows
```
Comment (jia200x) [inline sys/net/lora/Kconfig:117]: See https://github.com/RIOT-OS/RIOT/pull/15007/files#r497334915

File: sys/net/lora/Kconfig [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+config LORA_PAYLOAD_CRC_OFF_DEFAULT
+    bool "Disable payload CRC"
+
+config LORA_PAYLOAD_LENGTH_DEFAULT
+    int "Payload Length"
+    default 0
+    depends on !LORA_FIXED_HEADER_LEN_MODE_DEFAULT
+    help
+        Configure the length of payload. The option is disabled while using
```
Comment (jia200x) [inline sys/net/lora/Kconfig:153]: It's the other way around actually. This option is disabled when using explicit mode (since in that case the PHDR carries the length information).

Also, maybe this one shouldn't depend on `LORA_FIXED_HEADER_LEN_MODE_DEFAULT`, since there are use cases where this value never changes but the fixed header mode is enabled and disabled every now and then (e.g class B)

File: sys/net/lora/Kconfig [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+config LORA_SYMBOL_TIMEOUT_DEFAULT
+    int "Timeout for symbols in seconds"
+    default 10
+
+choice
+    bool "Channel bandwidth"
+    default LORA_BW_DEFAULT_125
+    help
+        Configure the channel bandwidth. Refer to country specific regulations
```
Comment (jia200x) [inline sys/net/lora/Kconfig:37]: I think the bandwidth is not region dependent (only the central frequency)

File: sys/net/lora/Kconfig [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        preamble. It starts with a series of upchirps to cover the whole
+        frequency band of the particular channel assigned. The last two upchirps
+        encode the sync word. Sync word is used to differntiate between LoRa
+        transmissions that use the same frequency bands. The sync word is
+        followed by two and a quarter downchirps, for a duration of 2.25
+        symbols. The total duration of this preamble can be configured between
+        10.25 and 65,539.25 symbol hence the range 8 to 65537.
+
+config LORA_SYMBOL_TIMEOUT_DEFAULT
+    int "Timeout for symbols in seconds"
```
Comment (jia200x) [inline sys/net/lora/Kconfig:30]: this symbol time out is in number of symbols, not in a time unit

File: sys/net/lora/Kconfig [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    bool "Enable Frequency hopping spread spectrum (FHSS)"
+    help
+        Enabling Frequency hopping spread spectrum (FHSS) would imply that a
+        portion of each LoRa packet is transmitted on each hopping channel from
+        a look up table of frequencies managed by the host microcontroller.
+
+config LORA_FREQUENCY_HOPPING_PERIOD_DEFAULT
+    int "Frequency hopping period"
+    default 0
+    depends on LORA_FREQUENCY_HOPPING_DEFAULT
```
Comment (jia200x) [inline sys/net/lora/Kconfig:134]: same as below, I think we don't need a dependency here. There are use cases where this value is set but the FHop is disabled by default.

File: sys/include/net/lora.h [hunk 10]
```diff
@@ -38,66 +40,161 @@ extern "C" {
...
+#define LORA_PAYLOAD_CRC_ON_DEFAULT                 (false)
+#else
+#define LORA_PAYLOAD_CRC_ON_DEFAULT                 (true)
+#endif
 #endif
 
+/** @brief Configure payload length
+ *
+ * Configure the length of payload. The mode is disabled when implicit header
+ * mode ( @ref CONFIG_LORA_FIXED_HEADER_LEN_MODE_DEFAULT ) is enabled.
```
Comment (jia200x) [inline sys/include/net/lora.h:193]: See https://github.com/RIOT-OS/RIOT/pull/15007/files#r497338460

File: sys/net/lora/Kconfig (hunk 7 above)
Comment (akshaim) [reply sys/net/lora/Kconfig:37]: In India it is. The carrier bandwidth is limited to 200KHz.

Refer : https://drive.google.com/file/d/1ECkRQ6yv11Kp5Ftz7oKtdgnI0K0uSA8G/view?usp=sharing

File: sys/net/lora/Kconfig (hunk 8 above)
Comment (akshaim) [reply sys/net/lora/Kconfig:30]: Oops yes.

File: sys/net/lora/Kconfig (hunk 9 above)
Comment (akshaim) [reply sys/net/lora/Kconfig:134]: Okay.  It looks obvious to me. Fine, I shall correct it.

File: sys/include/net/lora.h (hunk 4 above)
Comment (akshaim) [reply sys/include/net/lora.h:138]: Yes. Intially I wrote the help keeping in mind the LoRaWAN aspect. Shall correct it

File: sys/net/lora/Kconfig (hunk 7 above)
Comment (jia200x) [reply sys/net/lora/Kconfig:37]: ok, perfect then!

**Comment by akshaim** [comment]
I have removed all references tp LoRaWAN to avoid confusion (if any)

File: sys/include/net/lora.h [hunk 11]
```diff
@@ -38,66 +40,164 @@ extern "C" {
...
+#ifdef DOXYGEN
+#define CONFIG_LORA_FREQUENCY_HOPPING_DEFAULT
 #endif
 
-/** @brief Enable payload CRC, optional */
-#ifndef LORA_PAYLOAD_CRC_ON_DEFAULT
-#define LORA_PAYLOAD_CRC_ON_DEFAULT            (true)
+/** @brief Frequency hopping period
+ *
+ * Configure the hopping period, time which each transmission will dwell in any
```
Comment (jia200x) [inline sys/include/net/lora.h:161]: Could you add the unit here and in the Kconfig file? (unit is "symbols")

File: sys/include/net/lora.h [hunk 12]
```diff
@@ -38,66 +40,164 @@ extern "C" {
...
 
-/** @brief Preamble length, same for Tx and Rx */
-#ifndef LORA_PREAMBLE_LENGTH_DEFAULT
-#define LORA_PREAMBLE_LENGTH_DEFAULT           (8U)
+/** @brief Preamble length, same for Tx and Rx
+ *
+ * Configure preamble used in LoRa frame. Each LoRa frame begins with a
+ * preamble. It starts with a series of upchirps to cover the whole frequency
+ * band of the particular channel assigned. The last two upchirps encode the
+ * sync word. Sync word is used to differntiate between LoRa transmissions that
```
Comment (jia200x) [inline sys/include/net/lora.h:48]:
```suggestion
 * sync word. Sync word is used to differentiate between LoRa transmissions that
```

**Comment by jia200x** [comment]
I just added some last comment, but over all this looks ok. Please address those and also the doc issues pointed out by Travis-CI.

File: sys/include/net/lora.h (hunk 11 above)
Comment (akshaim) [reply sys/include/net/lora.h]: Done

**Comment by jia200x** [comment]
Looks good to me. Please squash

**Comment by jia200x** [comment]
BTW you can directly amend the doc issues when you squash:

```
	ERROR: Doxygen generates the following warnings:

	deprecated:32: warning: unable to resolve reference to `CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT' for \ref command

	sys/include/net/lora.h:180: warning: unable to resolve reference to `CONFIG_LORA_PAYLOAD_CRC_OFF_DEFAULT' for \ref 
```

**Comment by akshaim** [comment]
> Looks good to me. Please squash

Here we go ☺️

**Comment by jia200x** [comment]
great! let's see what Murdock says

**Review by jia200x — APPROVED** [review]
ACK

**Comment by jia200x** [comment]
Thanks for your contribution!
